hkmc2/shared/src/test/mlscript/handlers/UserThreads.mls

```diff
@@ -0,0 +1,88 @@
+:js
+:effectHandlers
+
+abstract class ThreadEffect with
+  val tasks = []
+  fun start(thread: ThreadEffect -> ()): ()
+  fun fork(thread: ThreadEffect -> ()): ()
+  fun yld(): ThreadEffect
+
+class Lock(locked) with
+  fun lock(th: ThreadEffect) =
+    while locked do
+      th.yld()
+    set locked = true
+  fun unlock() =
+    set locked = false
+
+fun f(x)(h) =
+  print("f " + x)
+  h.yld()
+  print("f " + x)
+  h.yld()
+  print("f " + x)
+
+fun main(h) =
+  print("main start")
+  h.fork(f(0))
+  h.fork(f(1))
+  h.fork(f(2))
+  print("main end")
+
+
+// LIFO
+handle h = ThreadEffect with
+  fun fork(thread)(k) =
+    this.tasks.unshift(thread)
+    k(())
+  fun yld()(k) =
+    this.tasks.push(k)
+    globalThis.eval("runtime").enterHandleBlock(this, () => this.tasks.shift()(this))
+  fun start(main)(k) =
+    this.tasks.push(main)
+    while this.tasks.length != 0 do
+      globalThis.eval("runtime").enterHandleBlock(this, () => this.tasks.shift()(this))
+    k()
+in
+  h.start(main)
```

Comment:
This is needed, because handler method themselves cannot throw effect of the handler itself. So here it just rebinds the handler and invoke it. Previously, the handler is still binded during the handler method with some weird semantics as the order of which handler is resumed is simply very messy (refer to `RecursiveHandler.mls` L56-79). This should be the correct behaviour now.

Reply:
Wait but that looks like major shift in the official semantics of effect handlers. I'm afraid it might make our support for the feature quite atypical/weird and not aligned with the standard definitions from the literature.

This seems like a significant design decision that should be properly discussed. Could you start by specifying what's the intended _semantics_ of `enterHandleBlock `? It doesn't seem to appear anywhere in your project report.

Reply:
Oh I just saw your other comment:

> The semantic of that function is simply entering a handle block with handler h binded, except as opposed to handle block, it does not create the handler h

1. Are you saying this is a way of obtaining deep handler semantics in an otherwise shallow handler implementation?

2. Could you please give examples of what happens if you forget to do it?

3. Why not implicitly wrap all handler blocks inside this call? Would that make sense?

Reply:
1. the handler is still deep, you can still raise effect multiple times in the handler body but not inside the actual handler. The actual handler method cannot raise effects.
2. This will be throwing error "Unhandled effect" as if the effect leaks.
3. Yes, this is possible. A handle block is simply `let h = new Handler(); runtime.enterHandleBlock(h, () => handlerBody);`

Reply:
So what happens here is this: because this.tasks contain some function that will still raise effect due to `.fork()` just adding those function directly, those added by `.yld` won't raise effects